File: transports/csharp/src/Csilgen.Transport/Conventions.cs
// Conventions shared by every CSIL transport -- see csil-transport-conventions.md.
//
// This file owns the parts the three transports agree on: the version constant, the transport
// status registry, tag-24 payload wrap/unwrap, the max-frame guard, the exception hierarchy, and
// the canonical-CBOR field accessors the envelopes build on so their bytes match the conformance
// vectors regardless of record layout.

namespace Csilgen.Transport;

// Status is a transport-level status. It is distinct from application errors, which ride inside the
// payload as a declared `/ ErrorType` arm. Equality is by the underlying code, so host-defined
// extension codes (>= 64) and unknown codes compare correctly. A struct because it is a single int.
public readonly record struct Status(int Code)
{
    public static readonly Status Ok = new(0);
    public static readonly Status MalformedEnvelope = new(1);
    public static readonly Status UnknownServiceOrOp = new(2);
    public static readonly Status Unauthenticated = new(3);
    public static readonly Status Forbidden = new(4);
    public static readonly Status VersionUnsupported = new(5);
    public static readonly Status Internal = new(6);
    public static readonly Status Unavailable = new(7);
    public static readonly Status DeadlineExceeded = new(8);

    // FromCode maps a wire code onto a Status, preserving host-defined and unknown codes verbatim.
    public static Status FromCode(int code) => new(code);

    // IsOk reports whether the status indicates a typed reply is present.
    public bool IsOk => Code == 0;

    // Name returns the registry name for the status, or "other" for codes outside it.
    public string Name => Code switch
    {
        0 => "ok",
        1 => "malformed-envelope",
        2 => "unknown-service-or-op",
        3 => "unauthenticated",
        4 => "forbidden",
        5 => "version-unsupported",
        6 => "internal",
        7 => "unavailable",
        8 => "deadline-exceeded",
        _ => "other",
    };
}

// TransportException is the base for errors surfaced by the transport layer (the wire), distinct
// from a host's application errors (which ride as a status-0 typed variant).
public class TransportException : Exception
{
    public TransportException(string message)
        : base(message)
    {
    }

    public TransportException(string message, Exception inner)
        : base(message, inner)
    {
    }
}

public sealed class EncodeException : TransportException
{
    public EncodeException(string message)
        : base(message)
    {
    }
}

public sealed class DecodeException : TransportException
{
    public DecodeException(string message)
        : base(message)
    {
    }
}

public sealed class MalformedException : TransportException
{
    public MalformedException(string message)
        : base("malformed envelope: " + message)
    {
    }
}

public sealed class FrameTooLargeException : TransportException
{
    public FrameTooLargeException(int got, int maximum)
        : base($"frame of {got} bytes exceeds max-frame guard of {maximum} bytes")
    {
        Got = got;
        Maximum = maximum;
    }

    public int Got { get; }

    public int Maximum { get; }
}

public sealed class UnsupportedVersionException : TransportException
{
    public UnsupportedVersionException(ulong version)
        : base($"unsupported transport version {version}")
    {
        Version = version;
    }

    public ulong Version { get; }
}

// StatusException carries a non-zero transport status returned by a peer, distinct from a host's
// application errors.
public sealed class StatusException : TransportException
{
    public StatusException(Status status, string? message)
        : base(message is null
            ? $"transport status {status.Name} ({status.Code})"
            : $"transport status {status.Name} ({status.Code}): {message}")
    {
        Status = status;
        StatusMessage = message;
    }

    public Status Status { get; }

    public string? StatusMessage { get; }
}

public sealed class CarrierException : TransportException
{
    public CarrierException(string message)
        : base("carrier error: " + message)
    {
    }
}

public static class Conventions
{
    // VERSION is the current transport version. A new value is minted only for a breaking change to
    // envelope layout or semantics.
    public const ulong Version = 1;

    // TagEncodedCbor is the CBOR semantic tag wrapping an embedded, opaque CBOR data item
    // (RFC 8949 §3.4.5.1).
    public const ulong TagEncodedCbor = 24;

    // ControlServiceOrd is the reserved service ordinal for the transport control plane (Events
    // lifecycle).
    public const ulong ControlServiceOrd = 0;

    // MaxFrameDefault is the default max encoded envelope size for stream/message carriers (16 MiB).
    // A carrier rejects anything larger before allocating for it.
    public const int MaxFrameDefault = 16 * 1024 * 1024;

    // Tag24 wraps opaque payload bytes (themselves a CBOR item) in tag 24.
    public static CborValue Tag24(byte[] payload) =>
        new CborTag(TagEncodedCbor, new CborBytes((byte[])payload.Clone()));

    // Untag24 extracts the opaque payload bytes from a tag-24 value.
    public static byte[] Untag24(CborValue value)
    {
        if (value is CborTag { Num: TagEncodedCbor, Content: CborBytes bytes })
        {
            return (byte[])bytes.Value.Clone();
        }

        throw new MalformedException("expected a tag-24 (encoded-cbor) payload");
    }

    // MapGet looks up a text key in a CBOR map value.
    public static CborValue? MapGet(CborValue value, string key)
    {
        if (value is CborMap map)
        {
            foreach (CborEntry entry in map.Entries)
            {
                if (entry.Key is CborText text && text.Value == key)
                {
                    return entry.Val;
                }
            }
        }

        return null;
    }

    // TryAsU64 reads a non-negative integer from a decoded CBOR integer value.
    public static bool TryAsU64(CborValue value, out ulong result)
    {
        switch (value)
        {
            case CborUint u:
                result = u.Value;
                return true;
            case CborInt i when i.Value >= 0:
                result = (ulong)i.Value;
                return true;
            default:
                result = 0;
                return false;
        }
    }

    // TryAsI64 reads a signed integer from a decoded CBOR integer value.
    public static bool TryAsI64(CborValue value, out long result)
    {
        switch (value)
        {
            case CborUint u:
                result = (long)u.Value;
                return true;
            case CborInt i:
                result = i.Value;
                return true;
            default:
                result = 0;
                return false;
        }
    }

    public static ulong GetUint(CborValue map, string key)
    {
        CborValue? v = MapGet(map, key);
        if (v is not null && TryAsU64(v, out ulong n))
        {
            return n;
        }

        throw new MalformedException($"missing or non-integer field '{key}'");
    }

    public static long GetInt(CborValue map, string key)
    {
        CborValue? v = MapGet(map, key);
        if (v is not null && TryAsI64(v, out long n))
        {
            return n;
        }

        throw new MalformedException($"missing or non-integer field '{key}'");
    }

    public static string GetText(CborValue map, string key)
    {
        CborValue? v = MapGet(map, key);
        if (v is CborText t)
        {
            return t.Value;
        }

        throw new MalformedException($"missing or non-text field '{key}'");
    }

    public static string? GetTextOpt(CborValue map, string key)
    {
        CborValue? v = MapGet(map, key);
        return v is CborText t ? t.Value : null;
    }

    public static ulong? GetUintOpt(CborValue map, string key)
    {
        CborValue? v = MapGet(map, key);
        if (v is not null && TryAsU64(v, out ulong n))
        {
            return n;
        }

        return null;
    }

    // CheckVersion verifies a decoded envelope's version field, throwing a clear error otherwise so
    // an unknown version is never silently misparsed.
    public static void CheckVersion(ulong version)
    {
        if (version != Version)
        {
            throw new UnsupportedVersionException(version);
        }
    }
}

File: transports/swift/Sources/CsilgenTransport/Conventions.swift
// Conventions shared by every CSIL transport — see csil-transport-conventions.md.
//
// This file owns the parts the three transports agree on: the version constant, the
// transport status registry, tag-24 payload wrap/unwrap, the max-frame guard, and the
// canonical-CBOR field accessors the envelopes build on so their bytes match the
// conformance vectors regardless of Swift stored-property order.

/// The current transport version. A new value is minted only for a breaking change to
/// envelope layout or semantics.
public let csilVersion: UInt64 = 1

/// The CBOR semantic tag wrapping an embedded, opaque CBOR data item (RFC 8949
/// §3.4.5.1).
public let tagEncodedCbor: UInt64 = 24

/// The reserved service ordinal for the transport control plane (Events lifecycle).
public let controlServiceOrd: UInt64 = 0

/// The default max encoded envelope size for stream/message carriers (16 MiB). A
/// carrier rejects anything larger before allocating for it.
public let maxFrameDefault: Int = 16 * 1024 * 1024

/// The conservative max datagram size (envelope + payload) safe across UDP/WebRTC/QUIC.
public let maxDatagramDefault: Int = 1200

/// Errors raised across the transport layer. Malformed wire input is *expected* and
/// must throw — never `fatalError`/`assert`, which are for programmer errors only.
public enum TransportError: Error, Equatable {
    case encode(String)
    case decode(String)
    case malformed(String)
    case frameTooLarge(got: Int, maximum: Int)
    case unsupportedVersion(UInt64)
    /// A non-zero transport status returned by a peer (distinct from an application
    /// error, which rides inside the payload as a declared `/ ErrorType` arm).
    case status(name: String, code: Int, message: String)
    case carrier(String)
}

/// A transport-level status. It is distinct from application errors. Modeled as a
/// `struct` over an `Int` (NOT a closed enum) so host-defined extension codes (>= 64)
/// and otherwise-unknown codes round-trip verbatim. Equality is by the code.
public struct Status: Equatable, Sendable {
    public let code: Int

    public init(code: Int) {
        self.code = code
    }

    // The registry codes (conventions doc §4).
    public static let ok = Status(code: 0)
    public static let malformedEnvelope = Status(code: 1)
    public static let unknownServiceOrOp = Status(code: 2)
    public static let unauthenticated = Status(code: 3)
    public static let forbidden = Status(code: 4)
    public static let versionUnsupported = Status(code: 5)
    public static let `internal` = Status(code: 6)
    public static let unavailable = Status(code: 7)
    public static let deadlineExceeded = Status(code: 8)

    /// Whether the status indicates a typed reply is present.
    public var isOk: Bool { code == 0 }

    /// The registry name for the status, or "other" for codes outside it.
    public var name: String {
        switch code {
        case 0: return "ok"
        case 1: return "malformed-envelope"
        case 2: return "unknown-service-or-op"
        case 3: return "unauthenticated"
        case 4: return "forbidden"
        case 5: return "version-unsupported"
        case 6: return "internal"
        case 7: return "unavailable"
        case 8: return "deadline-exceeded"
        default: return "other"
        }
    }
}

/// A malformed-envelope error with a formatted reason.
func malformed(_ reason: String) -> TransportError {
    TransportError.malformed(reason)
}

/// Wrap opaque payload bytes (themselves a CBOR item) in tag 24.
func tag24(_ payload: [UInt8]) -> CBORValue {
    .tag(tagEncodedCbor, .bytes(payload))
}

/// Extract the opaque payload bytes from a tag-24 value.
func untag24(_ value: CBORValue) throws -> [UInt8] {
    guard case .tag(let num, let content) = value, num == tagEncodedCbor else {
        throw malformed("expected a tag-24 (encoded-cbor) payload")
    }
    guard case .bytes(let b) = content else {
        throw malformed("tag-24 payload is not a byte string")
    }
    return b
}

/// Look up a text key in a CBOR map value.
func mapGet(_ value: CBORValue, _ key: String) -> CBORValue? {
    guard case .map(let entries) = value else {
        return nil
    }
    for entry in entries {
        if case .text(let t) = entry.key, t == key {
            return entry.value
        }
    }
    return nil
}

/// Read a non-negative integer from a decoded CBOR integer value.
func asU64(_ value: CBORValue) -> UInt64? {
    switch value {
    case .uint(let n): return n
    case .int(let n) where n >= 0: return UInt64(n)
    default: return nil
    }
}

/// Read a signed integer from a decoded CBOR integer value.
func asI64(_ value: CBORValue) -> Int64? {
    switch value {
    case .uint(let n) where n <= UInt64(Int64.max): return Int64(n)
    case .int(let n): return n
    default: return nil
    }
}

func getUint(_ map: CBORValue, _ key: String) throws -> UInt64 {
    guard let v = mapGet(map, key), let n = asU64(v) else {
        throw malformed("missing or non-integer field '\(key)'")
    }
    return n
}

func getInt(_ map: CBORValue, _ key: String) throws -> Int64 {
    guard let v = mapGet(map, key), let n = asI64(v) else {
        throw malformed("missing or non-integer field '\(key)'")
    }
    return n
}

func getText(_ map: CBORValue, _ key: String) throws -> String {
    guard let v = mapGet(map, key), case .text(let t) = v else {
        throw malformed("missing or non-text field '\(key)'")
    }
    return t
}

func getTextOpt(_ map: CBORValue, _ key: String) -> String? {
    guard let v = mapGet(map, key), case .text(let t) = v else {
        return nil
    }
    return t
}

func getUintOpt(_ map: CBORValue, _ key: String) -> UInt64? {
    guard let v = mapGet(map, key), let n = asU64(v) else {
        return nil
    }
    return n
}

/// Verify a decoded envelope's version field so an unknown version is never silently
/// misparsed.
func checkVersion(_ v: UInt64) throws {
    guard v == csilVersion else {
        throw TransportError.unsupportedVersion(v)
    }
}
